src/backend/macos/utils/user_alert.rs

```diff
@@ -0,0 +1,120 @@
+            response_flags.as_mut_ptr()
+        ) };
+        if is_canel != 0 {
+            return MessageDialogResult::Cancel;
+        }
+        let response = unsafe { response_flags.assume_init() };
+        match self.buttons {
+            MessageButtons::Ok if response == kCFUserNotificationDefaultResponse => MessageDialogResult::Ok,
+            MessageButtons::OkCancel if response == kCFUserNotificationDefaultResponse => MessageDialogResult::Ok,
+            MessageButtons::OkCancel if response == kCFUserNotificationAlternateResponse => MessageDialogResult::Cancel,
+            MessageButtons::YesNo if response == kCFUserNotificationDefaultResponse => MessageDialogResult::Yes,
+            MessageButtons::YesNo if response == kCFUserNotificationAlternateResponse => MessageDialogResult::No,
+            MessageButtons::YesNoCancel if response == kCFUserNotificationDefaultResponse => MessageDialogResult::Yes,
+            MessageButtons::YesNoCancel if response == kCFUserNotificationAlternateResponse => MessageDialogResult::No,
+            MessageButtons::YesNoCancel if response == kCFUserNotificationOtherResponse => MessageDialogResult::Cancel,
+            MessageButtons::OkCustom(custom) if response == kCFUserNotificationDefaultResponse => MessageDialogResult::Custom(custom.to_owned()),
+            MessageButtons::OkCancelCustom(custom, _) if response == kCFUserNotificationDefaultResponse => MessageDialogResult::Custom(custom.to_owned()),
+            MessageButtons::OkCancelCustom(_, custom) if response == kCFUserNotificationAlternateResponse => MessageDialogResult::Custom(custom.to_owned()),
+            MessageButtons::YesNoCancelCustom(custom, _, _) if response == kCFUserNotificationDefaultResponse => MessageDialogResult::Custom(custom.to_owned()),
+            MessageButtons::YesNoCancelCustom(_, custom, _) if response == kCFUserNotificationAlternateResponse => MessageDialogResult::Custom(custom.to_owned()),
+            MessageButtons::YesNoCancelCustom(_, _, custom) if response == kCFUserNotificationOtherResponse => MessageDialogResult::Custom(custom.to_owned()),
+            _ => MessageDialogResult::Cancel,
+        }
+    }
+}
+pub fn sync_pop_dialog(opt: MessageDialog, mtm: MainThreadMarker) -> MessageDialogResult {
+    UserAlert::new(opt, Some(mtm)).run()
+}
+pub fn async_pop_dialog(opt: MessageDialog) -> DialogFutureType<MessageDialogResult> {
+    let deferred_future = ThreadDeferredFuture::default();
+    let defer = deferred_future.defer();
+    let opt2 = opt.clone();
+    let result: Result<(), String> = ThreadPool::new().map_err(|err| err.to_string()).and_then(|thread_pool| thread_pool.spawn(async move {
+        let mut defer = defer.lock().unwrap_or_else(PoisonError::into_inner);
+        let message_dialog_result = UserAlert::new(opt2.clone(), None).run();
+        defer.complete(message_dialog_result);
+    }).map_err(|err| err.to_string()));
```

Comment:
Why do we need a whole futures crate, and it's ThreadPool just to spawn a single thread, and discard the pool every time?

Looks like this should be a simple `std::thread::spawn(|| {})`

Also, are you sure that `CFUserNotificationDisplayAlert` is thread safe? Usually, macOS's UI related functions are not supposed to be called from non-main threads. Perhaps CF is smarter and allows this, but I'm not able to find any information about this.

Reply:
@PolyMeilex You're right. Accordingly, I have just submitted another [git commit](https://github.com/stuartZhang/rfd/commit/50326a748b6ddff7a4c9d94fe888bfee5042ffd3) to remedy them:

1. Remove the conditional dependency on the `futures` crate.
2. Relinquish the heavy thread-pool instrument.
3. Spawn a background thread by means of the std lib.